File: SDS/time/time_B_2805.cpp
#include<stdio.h>
#include<algorithm>

using namespace std;

int ans;
int height[1000000];

bool desc(int a, int b)
{
    return a > b;
}

int main()
{
    int N, M;
    int MAX = 0;
    
    scanf("%d %d", &N, &M);

    for(int i=0; i<N; i++)
    {
        scanf("%d", &height[i]);
        MAX += height[i];
    }
    
    sort(height, height+N,  desc);
    
    int low = 0;
    int high = height[0];
    int mid = 0;
    
    while(low <= high)
    {
        int total = MAX;
        
        mid = (low + high) >> 1;
        
        for(int i =0; i< N; i++)
        {
            if(height[i] >= mid)
                total -= mid;
            else
                total -= height[i];
        }
        
        if(total == M)
        {
            ans = mid;
            break;
        }
        
        if(M < total)
            low = mid + 1;
        else
        {
            ans = mid;
            high = mid - 1;
        }
    }
    
    if(ans == -1)
        ans = (low + high) >> 1;
    
    printf("%d", ans);
}

File: SDS/time/time_B_2805.py
N, M = map(int, input().split())
height = sorted([int(x) for x in input().split()], reverse=True)
MAX = sum(height)

low = 0
high = height[0]
ans = 0
find = False

while low <= high:
    total = MAX
    mid = (low + high) >> 1
    
    for i in range(N):
        if height[i] >= mid:
            total -= mid
        else:
            total -= height[i]
    
    if total == M:
        ans = mid
        find = True
        
    if M < total:
        low = mid +1
    else:
        ans = mid
        high = mid - 1
        
if not find:
    ans = (low + high) >> 1
        
print(ans)
